R/s3.R

```diff
@@ -0,0 +1,163 @@
+}
+
+
+##' @export
+##' @rdname storr_s3
+driver_s3 <- function(bucket, remote_root, ..., path_local = NULL) {
+  ops <- s3_file_ops(bucket, remote_root)
+  storr::driver_remote(ops, ..., path_local = path_local)
+}
+
+
+s3_file_ops <- function(bucket, root) {
+  R6_s3_file_ops$new(bucket, root)
+}
+
+
+R6_s3_file_ops <- R6::R6Class(
+  "s3_file_ops",
+
+  public = list(
+    bucket = NULL,
+
+    initialize = function(bucket, root) {
+      self$bucket <- bucket
+      self$root <- root
+    },
+
+    type = function() {
+      "s3"
+    },
+
+    destroy = function() {
+      ## TODO: not sure if this is right?
+      self$file_ops$delete_dir(path = self$root)
+    },
+
+    create_dir = function(path) {
+      path_remote <- file.path(self$root, path)
+      aws.s3::put_folder(folder = path_remote, bucket = self$bucket)
+    },
+
+    list_dir = function(path) {
+      if (substr(path, nchar(path), nchar(path)) != "/") {
+        path <- paste0(path, "/")
+      }
+      path_remote <- file.path(self$root)
+      files_table <- aws.s3::get_bucket_df(bucket = self$bucket,
+                                           prefix = path_root,
+                                           max = Inf)
+      keys <- files_table[files_table$Size > 0,]$Key
+      ## TODO: this should really be stripped off the front rather
+      ## than gsub?  And probably fixed replacement to avoid
+      ## regexp-sensitive characters messing things up?
+      files <- gsub(pattern = path_root, replacement = "", x = keys)
+      split_names <- strsplit(files, "/")
+      ## first element of each split name is the file or directory
+      ## within path, take unique of these, so that directories only
+      ## appear once
+      ##
+      ## TODO: I _think_ I've done the translation from
+      ## unique/unlist/lapply to unique/vapply here, but I don't
+      ## really get what is going on.  It's quite likely that what you
+      ## really want is
+      ##
+      ##   unique(basename(files))
+      ##
+      ## perhaps?
+      unique(vapply(split_names, function(x) x[[1L]], character(1)))
+    },
+
+    exists = function(path) {
+      path_remote <- file.path(self$root, path)
+      ## QUESTION: What messages are produced by this?
+      suppressMessages(
+        aws.s3::head_object(object = path_remote, bucket = self$bucket)[[1]])
+    },
+
+    exists_dir = function(path) {
+      ## TODO: If it is possible to test if the remote thing is a
+      ## directory (e.g., one of the bits of head_object?) that would
+      ## be nice to get here too...
+      self$exists(path)
+    },
+
+    delete_file = function(path) {
+      if (length(path) != 1L) {
+        ## TODO: this is the most naive way of doing vectorisation,
+        ## but if aws.s3 can delete multiple objects (the docs suggest
+        ## it can) you can probably do better.  This function must
+        ## return a logical vector the same length as "path"
+        ## indicating if deletion actually happened.
```

Comment:
seems like a good way to have the functionality as a stop gap - I will try to take a look at this when I get a chance.